### ISSUE TITLE
Non-uniform timestamp format between querying and writing

### ISSUE THREAD

**Issue by dandv** [node-influx/node-influx#203]
I was trying to use queryRaw to do some data migration, and found that a point retrieved via `.queryRaw` isn't easily passed to `.writePoint`.

`queryRaw` (and `query` as well) pass through the JSON returned by Influx, which [serializes dates into RFC3339 strings](https://docs.influxdata.com/influxdb/v1.0/guides/querying_data/#querying-data-using-the-http-api).

The problem is that `.writePoints` expects a [Date object or a UNIX time](https://github.com/node-influx/node-influx/blob/ba63edf83f1ecd9e09e4814c6d6f6626962c4c98/index.js#L399).

**Comment by dandv** [comment]
Some docs on the (non)idempotence of timestamps would be useful.

**Comment by connor4312** [comment]
Reopening to remind myself

**Comment by dandv** [comment]
Also good to know about the [time format in responses](https://docs.influxdata.com/influxdb/v1.0/guides/querying_data/#other-options-when-querying-data), since this may help optimize parsing:

> Everything in InfluxDB is stored and reported in UTC. By default, timestamps are returned in RFC3339 UTC and have nanosecond precision, for example 2015-08-04T19:05:14.318570484Z. If you want timestamps in Unix epoch format include in your request the query string parameter `epoch` where `epoch=[h,m,s,ms,u,ns]`. For example, get epoch in seconds with:

    curl -G 'http://localhost:8086/query' --data-urlencode "db=mydb" --data-urlencode "epoch=s" --data-urlencode "q=SELECT \"value\" FROM \"cpu_load_short\" WHERE \"region\"='us-west'"

**Comment by connor4312** [comment]
So, here's my [big scary implementation of time handling](https://github.com/node-influx/node-influx/blob/5.0.0/src/grammar/times.ts#L1). The comment in there explains it decently enough. It is by no means super fun code, but it seems to work.

In effect this allows idempotent reading and writing from Influx. The dates that are returned from Influx are, when precision is unspecified (ns) or set to a greater-than-millisecond precision, decorated with two methods which allow the consumer to retrieve the nanosecond unix timestamp and the nanosecond RFC3339 timestamp.

```ts
export interface NanoDate extends Date {
  /**
   * Returns the unix nanoseconds timestamp as a string.
   */
  getNanoTime(): string;

  /**
   * Formats the date as an ISO RFC3339 timestamp with nanosecond precision.
   */
  toNanoISOString(): string;
}
```

When sending data to Influx, we check to see if the provided dates are NanoDates and, if so, we represent them correctly, either in Influx's custom time format when inserting them into queries, or in nanoseconds timestamps in line protocol.

If the consumer wants to write a date where they have a nanosecond timestamp that they retrieved outside of Influx, we export a `toNanoDate` function for that purpose.

```ts
/**
 * Covers a nanoseconds unix timestamp to a NanoDate for node-influx. The
 * timestamp is provided as a string to prevent precision loss.
 */
export function toNanoDate(timestamp: string): NanoDate;
```

---

I think that should above cover it. It's a bit of an awkward dance between three formats in a language that can't correctly represent one of them. (At least Influx doesn't support non-UTC timezones... yet.) Feel free to post feedback, as always.

**Comment by connor4312** [comment]
Fyi wrote up a page on the docs with a bit more friendly, high-level explanation on how times are dealt with: https://node-influx.github.io/manual/usage.html#a-moment-for-times

**Comment by connor4312** [comment]
These changes have been published in `5.0.0-alpha.1`. You can install this via:

```
npm install --save influx@next
```

**Comment by SamAlbuga** [comment]
I still can't get node-influx to return the time like the format you would get in a request with `epoch=s`.
Can anyone post an example?? I'm using an iOS charts library that just requires basic timestamps in seconds!
